python/ray/air/_internal/checkpoint_manager.py

```diff
@@ -382,15 +382,20 @@ def _get_checkpoint_score(
         checkpoint_score_attribute = (
             self._checkpoint_strategy.checkpoint_score_attribute
         )
-        if checkpoint_score_attribute not in checkpoint.metrics:
-            logger.error(
-                f"Result dict has no key: {checkpoint_score_attribute}. "
-                f"checkpoint_score_attr must be set to a key in the "
-                f"result dict. Valid keys are: {list(checkpoint.metrics.keys())}"
-            )
-            checkpoint_result = float("-inf")
+        if checkpoint_score_attribute:
+            flat_metrics = flatten_dict(checkpoint.metrics)
+            try:
+                checkpoint_result = flat_metrics[checkpoint_score_attribute]
+            except KeyError:
+                valid_keys = list(flat_metrics.keys())
+                logger.error(
+                    f"Result dict has no key: {checkpoint_score_attribute}. "
+                    f"checkpoint_score_attr must be set to a key in the "
+                    f"result dict. Valid keys are: {valid_keys}"
+                )
+                checkpoint_result = float("-inf")
         else:
-            checkpoint_result = checkpoint.metrics[checkpoint_score_attribute]
+            checkpoint_result = float("-inf")
```

Comment:
Should the default result should be the training iteration? We document that the earliest checkpoints get deleted if no `checkpoint_score_attribute` is given, but I think that just happens implicitly through the FIFO heap push/pop order -- maybe we should just always order the heap?

Reply:
I've added a docstring that may shed some light on this. We actually explicitly order by checkpoint ID in the last tuple entry. The logic is as follows:

`_get_checkpoint_score` returns a tuple `(is_not_na, metric, id)`

Higher values are preferred/kept.

`sort()` first sorts by the first tuple index, then the second, then the third.

The first index ensures that NaN metrics are discarded first (rated worst).

The second index then sorts by actual metric.

If the metrics are the same (e.g. when no attr is defined --> we get -inf), then we sort by checkpoint ID and keep the later checkpoints.

Reply:
Got it, that clears it up a lot!